Jobs often fail with Galaxy giving 500s

This is a failure that can happen early:

```
Run sudo --preserve-env=ANSIBLE_PYTHON_INTERPRETER,ATMOSPHERE_NETWORK_BACKEND,MOLECULE_CSI_DRIVER env "PATH=$PATH" molecule test -s aio
WARNING  Molecule scenarios should migrate to 'extensions/molecule'
INFO     [aio > discovery] scenario test matrix: dependency, cleanup, destroy, syntax, create, prepare, converge, idempotence, side_effect, verify, cleanup, destroy
INFO     [aio > prerun] Performing prerun with role_name_check=0...
WARNING  Retrying execution failure 250 of: ansible-galaxy collection install -vvv community.crypto:2.2.3
ERROR    Command ansible-galaxy collection install -vvv community.crypto:2.2.3, returned 250 code:
Starting galaxy collection install process
Process install dependency map
the full traceback was:

Traceback (most recent call last):
  File "/opt/github/actions-runner/2.330.0/_work/atmosphere/atmosphere/.venv/lib/python3.10/site-packages/ansible/cli/__init__.py", line 658, in cli_executor
    exit_code = cli.run()
  File "/opt/github/actions-runner/2.330.0/_work/atmosphere/atmosphere/.venv/lib/python3.10/site-packages/ansible/cli/galaxy.py", line 773, in run
    return context.CLIARGS()
  File "/opt/github/actions-runner/2.330.0/_work/atmosphere/atmosphere/.venv/lib/python3.10/site-packages/ansible/cli/galaxy.py", line 120, in method_wrapper
    return wrapped_method(*args, **kwargs)
  File "/opt/github/actions-runner/2.330.0/_work/atmosphere/atmosphere/.venv/lib/python3.10/site-packages/ansible/cli/galaxy.py", line 1438, in execute_install
    self._execute_install_collection(
  File "/opt/github/actions-runner/2.330.0/_work/atmosphere/atmosphere/.venv/lib/python3.10/site-packages/ansible/cli/galaxy.py", line 1487, in _execute_install_collection
    install_collections(
  File "/opt/github/actions-runner/2.330.0/_work/atmosphere/atmosphere/.venv/lib/python3.10/site-packages/ansible/galaxy/collection/__init__.py", line 735, in install_collections
    dependency_map = _resolve_depenency_map(
  File "/opt/github/actions-runner/2.330.0/_work/atmosphere/atmosphere/.venv/lib/python3.10/site-packages/ansible/galaxy/collection/__init__.py", line 1852, in _resolve_depenency_map
    return collection_dep_resolver.resolve(
  File "/opt/github/actions-runner/2.330.0/_work/atmosphere/atmosphere/.venv/lib/python3.10/site-packages/resolvelib/resolvers.py", line 546, in resolve
    state = resolution.resolve(requirements, max_rounds=max_rounds)
  File "/opt/github/actions-runner/2.330.0/_work/atmosphere/atmosphere/.venv/lib/python3.10/site-packages/resolvelib/resolvers.py", line 397, in resolve
    self._add_to_criteria(self.state.criteria, r, parent=None)
  File "/opt/github/actions-runner/2.330.0/_work/atmosphere/atmosphere/.venv/lib/python3.10/site-packages/resolvelib/resolvers.py", line 148, in _add_to_criteria
    matches = self._p.find_matches(
  File "/opt/github/actions-runner/2.330.0/_work/atmosphere/atmosphere/.venv/lib/python3.10/site-packages/ansible/galaxy/dependency_resolution/providers.py", line 477, in find_matches
    match for match in self._find_matches(list(requirements))
  File "/opt/github/actions-runner/2.330.0/_work/atmosphere/atmosphere/.venv/lib/python3.10/site-packages/ansible/galaxy/dependency_resolution/providers.py", line 253, in _find_matches
    coll_versions =  if preinstalled_candidates else self._api_proxy.get_collection_versions(first_req)  # type: t.Iterable]
  File "/opt/github/actions-runner/2.330.0/_work/atmosphere/atmosphere/.venv/lib/python3.10/site-packages/ansible/galaxy/collection/galaxy_api_proxy.py", line 106, in get_collection_versions
    return set(
  File "/opt/github/actions-runner/2.330.0/_work/atmosphere/atmosphere/.venv/lib/python3.10/site-packages/ansible/galaxy/collection/galaxy_api_proxy.py", line 106, in <genexpr>
    return set(
  File "/opt/github/actions-runner/2.330.0/_work/atmosphere/atmosphere/.venv/lib/python3.10/site-packages/ansible/galaxy/collection/galaxy_api_proxy.py", line 87, in _get_collection_versions
    raise last_error
  File "/opt/github/actions-runner/2.330.0/_work/atmosphere/atmosphere/.venv/lib/python3.10/site-packages/ansible/galaxy/collection/galaxy_api_proxy.py", line 66, in _get_collection_versions
    versions = api.get_collection_versions(requirement.namespace, requirement.name)
  File "/opt/github/actions-runner/2.330.0/_work/atmosphere/atmosphere/.venv/lib/python3.10/site-packages/ansible/galaxy/api.py", line 145, in wrapped
    return method(self, *args, **kwargs)
  File "/opt/github/actions-runner/2.330.0/_work/atmosphere/atmosphere/.venv/lib/python3.10/site-packages/ansible/galaxy/api.py", line 874, in get_collection_versions
    data = self._call_galaxy(versions_url, error_context_msg=error_context_msg, cache=True, cache_key=cache_key)
  File "/opt/github/actions-runner/2.330.0/_work/atmosphere/atmosphere/.venv/lib/python3.10/site-packages/ansible/module_utils/api.py", line 169, in run_function
    return call_retryable_function()
  File "/opt/github/actions-runner/2.330.0/_work/atmosphere/atmosphere/.venv/lib/python3.10/site-packages/ansible/galaxy/api.py", line 390, in _call_galaxy
    res = path_cache
KeyError: 'results'

ansible-galaxy 
  config file = None
  configured module search path = 
  ansible python module location = /opt/github/actions-runner/2.330.0/_work/atmosphere/atmosphere/.venv/lib/python3.10/site-packages/ansible
  ansible collection location = /root/.ansible/collections:/usr/share/ansible/collections:/opt/github/actions-runner/2.330.0/_work/atmosphere/atmosphere/.venv/lib/python3.10/dist-packages:/opt/github/actions-runner/2.330.0/_work/atmosphere/atmosphere/.venv/lib/python3/dist-packages:/opt/github/actions-runner/2.330.0/_work/atmosphere/atmosphere/.venv/local/lib/python3.10/dist-packages:/opt/github/actions-runner/2.330.0/_work/atmosphere/atmosphere/.venv/lib/python3.10/site-packages
  executable location = /opt/github/actions-runner/2.330.0/_work/atmosphere/atmosphere/.venv/bin/ansible-galaxy
  python version = 3.10.12 (main, Feb  4 2025, 14:57:36)  (/opt/github/actions-runner/2.330.0/_work/atmosphere/atmosphere/.venv/bin/python)
  jinja version = 3.1.6
  libyaml = True
No config file found; using defaults
: Collection ansible.netcommon does not support Ansible version
Traceback (most recent call last):
  File "/opt/github/actions-runner/2.330.0/_work/atmosphere/atmosphere/.venv/bin/molecule", line 10, in <module>
2.17.12
Found installed collection ansible.netcommon:1.2.0 at '/root/.ansible/collections/ansible_collections/ansible/netcommon'
Found installed collection ansible.utils:6.0.1 at '/root/.ansible/collections/ansible_collections/ansible/utils'
Found installed collection ansible.posix:1.6.0 at '/root/.ansible/collections/ansible_collections/ansible/posix'
Opened /root/.ansible/galaxy_token
: Skipping Galaxy server [https://galaxy.ansible.com/api/.](https://galaxy.ansible.com/api/) Got an
unexpected error when getting available versions of collection
community.crypto: 'results'
ERROR! Unexpected Exception, this is probably a bug: 'results'

    sys.exit(main())
  File "/opt/github/actions-runner/2.330.0/_work/atmosphere/atmosphere/.venv/lib/python3.10/site-packages/click/core.py", line 1442, in __call__
    return self.main(*args, **kwargs)
  File "/opt/github/actions-runner/2.330.0/_work/atmosphere/atmosphere/.venv/lib/python3.10/site-packages/click/core.py", line 1363, in main
    rv = self.invoke(ctx)
  File "/opt/github/actions-runner/2.330.0/_work/atmosphere/atmosphere/.venv/lib/python3.10/site-packages/click/core.py", line 1830, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/opt/github/actions-runner/2.330.0/_work/atmosphere/atmosphere/.venv/lib/python3.10/site-packages/click/core.py", line 1226, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/opt/github/actions-runner/2.330.0/_work/atmosphere/atmosphere/.venv/lib/python3.10/site-packages/click/core.py", line 794, in invoke
    return callback(*args, **kwargs)
  File "/opt/github/actions-runner/2.330.0/_work/atmosphere/atmosphere/.venv/lib/python3.10/site-packages/molecule/click_cfg.py", line 583, in wrapper
    return func(*args, **kwargs)
  File "/opt/github/actions-runner/2.330.0/_work/atmosphere/atmosphere/.venv/lib/python3.10/site-packages/click/decorators.py", line 34, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/opt/github/actions-runner/2.330.0/_work/atmosphere/atmosphere/.venv/lib/python3.10/site-packages/molecule/click_cfg.py", line 418, in wrapper
    return func(ctx)
  File "/opt/github/actions-runner/2.330.0/_work/atmosphere/atmosphere/.venv/lib/python3.10/site-packages/molecule/command/test.py", line 81, in test
    base.execute_cmdline_scenarios(scenario_name, args, command_args, ansible_args, exclude)
  File "/opt/github/actions-runner/2.330.0/_work/atmosphere/atmosphere/.venv/lib/python3.10/site-packages/molecule/command/base.py", line 169, in execute_cmdline_scenarios
    _run_scenarios(scenarios, command_args, default_config)
  File "/opt/github/actions-runner/2.330.0/_work/atmosphere/atmosphere/.venv/lib/python3.10/site-packages/molecule/command/base.py", line 240, in _run_scenarios
    scenario.config.runtime.prepare_environment(
  File "/opt/github/actions-runner/2.330.0/_work/atmosphere/atmosphere/.venv/lib/python3.10/site-packages/ansible_compat/runtime.py", line 694, in prepare_environment
    self.install_collection(
  File "/opt/github/actions-runner/2.330.0/_work/atmosphere/atmosphere/.venv/lib/python3.10/site-packages/ansible_compat/runtime.py", line 552, in install_collection
    raise InvalidPrerequisiteError(msg)
ansible_compat.errors.InvalidPrerequisiteError: Command ansible-galaxy collection install -vvv community.crypto:2.2.3, returned 250 code:
Starting galaxy collection install process
Process install dependency map
the full traceback was:

Traceback (most recent call last):
  File "/opt/github/actions-runner/2.330.0/_work/atmosphere/atmosphere/.venv/lib/python3.10/site-packages/ansible/cli/__init__.py", line 658, in cli_executor
    exit_code = cli.run()
  File "/opt/github/actions-runner/2.330.0/_work/atmosphere/atmosphere/.venv/lib/python3.10/site-packages/ansible/cli/galaxy.py", line 773, in run
    return context.CLIARGS['func']()
  File "/opt/github/actions-runner/2.330.0/_work/atmosphere/atmosphere/.venv/lib/python3.10/site-packages/ansible/cli/galaxy.py", line 120, in method_wrapper
    return wrapped_method(*args, **kwargs)
  File "/opt/github/actions-runner/2.330.0/_work/atmosphere/atmosphere/.venv/lib/python3.10/site-packages/ansible/cli/galaxy.py", line 1438, in execute_install
    self._execute_install_collection(
  File "/opt/github/actions-runner/2.330.0/_work/atmosphere/atmosphere/.venv/lib/python3.10/site-packages/ansible/cli/galaxy.py", line 1487, in _execute_install_collection
    install_collections(
  File "/opt/github/actions-runner/2.330.0/_work/atmosphere/atmosphere/.venv/lib/python3.10/site-packages/ansible/galaxy/collection/__init__.py", line 735, in install_collections
    dependency_map = _resolve_depenency_map(
  File "/opt/github/actions-runner/2.330.0/_work/atmosphere/atmosphere/.venv/lib/python3.10/site-packages/ansible/galaxy/collection/__init__.py", line 1852, in _resolve_depenency_map
    return collection_dep_resolver.resolve(
  File "/opt/github/actions-runner/2.330.0/_work/atmosphere/atmosphere/.venv/lib/python3.10/site-packages/resolvelib/resolvers.py", line 546, in resolve
    state = resolution.resolve(requirements, max_rounds=max_rounds)
  File "/opt/github/actions-runner/2.330.0/_work/atmosphere/atmosphere/.venv/lib/python3.10/site-packages/resolvelib/resolvers.py", line 397, in resolve
    self._add_to_criteria(self.state.criteria, r, parent=None)
  File "/opt/github/actions-runner/2.330.0/_work/atmosphere/atmosphere/.venv/lib/python3.10/site-packages/resolvelib/resolvers.py", line 148, in _add_to_criteria
    matches = self._p.find_matches(
  File "/opt/github/actions-runner/2.330.0/_work/atmosphere/atmosphere/.venv/lib/python3.10/site-packages/ansible/galaxy/dependency_resolution/providers.py", line 477, in find_matches
    match for match in self._find_matches(list(requirements[identifier]))
  File "/opt/github/actions-runner/2.330.0/_work/atmosphere/atmosphere/.venv/lib/python3.10/site-packages/ansible/galaxy/dependency_resolution/providers.py", line 253, in _find_matches
    coll_versions = [] if preinstalled_candidates else self._api_proxy.get_collection_versions(first_req)  # type: t.Iterable[t.Tuple[str, GalaxyAPI]]
  File "/opt/github/actions-runner/2.330.0/_work/atmosphere/atmosphere/.venv/lib/python3.10/site-packages/ansible/galaxy/collection/galaxy_api_proxy.py", line 106, in get_collection_versions
    return set(
  File "/opt/github/actions-runner/2.330.0/_work/atmosphere/atmosphere/.venv/lib/python3.10/site-packages/ansible/galaxy/collection/galaxy_api_proxy.py", line 106, in <genexpr>
    return set(
  File "/opt/github/actions-runner/2.330.0/_work/atmosphere/atmosphere/.venv/lib/python3.10/site-packages/ansible/galaxy/collection/galaxy_api_proxy.py", line 87, in _get_collection_versions
    raise last_error
  File "/opt/github/actions-runner/2.330.0/_work/atmosphere/atmosphere/.venv/lib/python3.10/site-packages/ansible/galaxy/collection/galaxy_api_proxy.py", line 66, in _get_collection_versions
    versions = api.get_collection_versions(requirement.namespace, requirement.name)
  File "/opt/github/actions-runner/2.330.0/_work/atmosphere/atmosphere/.venv/lib/python3.10/site-packages/ansible/galaxy/api.py", line 145, in wrapped
    return method(self, *args, **kwargs)
  File "/opt/github/actions-runner/2.330.0/_work/atmosphere/atmosphere/.venv/lib/python3.10/site-packages/ansible/galaxy/api.py", line 874, in get_collection_versions
    data = self._call_galaxy(versions_url, error_context_msg=error_context_msg, cache=True, cache_key=cache_key)
  File "/opt/github/actions-runner/2.330.0/_work/atmosphere/atmosphere/.venv/lib/python3.10/site-packages/ansible/module_utils/api.py", line 169, in run_function
    return call_retryable_function()
  File "/opt/github/actions-runner/2.330.0/_work/atmosphere/atmosphere/.venv/lib/python3.10/site-packages/ansible/galaxy/api.py", line 390, in _call_galaxy
    res = path_cache['results']
KeyError: 'results'

ansible-galaxy [core 2.17.12]
  config file = None
  configured module search path = ['/root/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /opt/github/actions-runner/2.330.0/_work/atmosphere/atmosphere/.venv/lib/python3.10/site-packages/ansible
  ansible collection location = /root/.ansible/collections:/usr/share/ansible/collections:/opt/github/actions-runner/2.330.0/_work/atmosphere/atmosphere/.venv/lib/python3.10/dist-packages:/opt/github/actions-runner/2.330.0/_work/atmosphere/atmosphere/.venv/lib/python3/dist-packages:/opt/github/actions-runner/2.330.0/_work/atmosphere/atmosphere/.venv/local/lib/python3.10/dist-packages:/opt/github/actions-runner/2.330.0/_work/atmosphere/atmosphere/.venv/lib/python3.10/site-packages
  executable location = /opt/github/actions-runner/2.330.0/_work/atmosphere/atmosphere/.venv/bin/ansible-galaxy
  python version = 3.10.12 (main, Feb  4 2025, 14:57:36) [GCC 11.4.0] (/opt/github/actions-runner/2.330.0/_work/atmosphere/atmosphere/.venv/bin/python)
  jinja version = 3.1.6
  libyaml = True
No config file found; using defaults
Warning: : Collection ansible.netcommon does not support Ansible version
2.17.12
Found installed collection ansible.netcommon:1.2.0 at '/root/.ansible/collections/ansible_collections/ansible/netcommon'
Found installed collection ansible.utils:6.0.1 at '/root/.ansible/collections/ansible_collections/ansible/utils'
Found installed collection ansible.posix:1.6.0 at '/root/.ansible/collections/ansible_collections/ansible/posix'
Opened /root/.ansible/galaxy_token
Warning: : Skipping Galaxy server [https://galaxy.ansible.com/api/.](https://galaxy.ansible.com/api/) Got an
unexpected error when getting available versions of collection
community.crypto: 'results'
ERROR! Unexpected Exception, this is probably a bug: 'results'

Error: Process completed with exit code 1.
```

We need to figure out a way to retry for this.

--- a/.zuul.yaml
+++ b/.zuul.yaml
@@ -28,6 +28,19 @@
     nodeset: atmosphere-ubuntu-jammy-single-node
     pre-run: test-playbooks/molecule/pre.yml
     post-run: test-playbooks/molecule/post.yml
+    required-projects:
+      - ansible-collections/ansible.netcommon
+      - ansible-collections/ansible.posix
+      - ansible-collections/ansible.utils
+      - ansible-collections/community.crypto
+      - ansible-collections/community.general
+      - ansible-collections/community.mysql
+      - ansible-collections/kubernetes.core
+      - openstack/ansible-collections-openstack
+      - vexxhost/ansible-collection-ceph
+      - vexxhost/ansible-collection-containers
+      - vexxhost/ansible-collection-kubernetes
+      - vexxhost/atmosphere.common
     roles:
       - zuul: opendev.org/openstack/openstack-helm
     vars:

--- a/releasenotes/notes/use-zuul-checkouts-for-molecule-collections-c2ede341.yaml
+++ b/releasenotes/notes/use-zuul-checkouts-for-molecule-collections-c2ede341.yaml
@@ -1,0 +1,6 @@
+---
+fixes:
+  - |
+    Molecule CI jobs now use checked-out sibling collection projects for
+    Atmosphere's Galaxy dependencies, avoiding external Galaxy downloads for
+    those collections during setup.

